mib-generator/cw_mib_generator.py

```diff
@@ -0,0 +1,145 @@
+
+    return mib_data
+
+
+def write_mib_file(mib_path, mib_contents):
+    # Try to write specified data to file at specified MIB path.
+    try:
+        with open(mib_path, "w") as mib_file:
+            mib_file.write(mib_contents)
+    except IOError as ioe:
+        print_err_and_exit("Could not write MIB file:\n - {}".format(ioe))
+
+
+def generate_title(mib_file_path):
+    # Try to extract MIB name from the file path, and append definitions
+    # statement.
+    try:
+        name = re.search(r"[\w\-]+$", mib_file_path).group(0)
+    except AttributeError:
+        print_err_and_exit("Could not find a valid MIB file name in: '{}'"
+                           .format(mib_file_path))
+
+    return name + " DEFINITIONS ::= BEGIN"
+
+
+def generate_mib(extras_file_path, output_mib_path):
+    # Read in the common and specific MIB fragments, fill in templated lines
+    # and write the complete MIB file.
+    full_common_mib_path = os.path.abspath(COMMON_MIB_PATH)
+    raw_common_data = read_mib_fragment(full_common_mib_path)
+    common_src_template = Template(raw_common_data)
+
+    substitute_dict = { 'title_statement' : generate_title(output_mib_path),
+                        'direct_edit_statement' : EDIT_STATEMENT }
+
+    try:
+        common_src = common_src_template.substitute(substitute_dict)
+    except ValueError as ve:
+        print_err_and_exit(
+            "Common MIB fragment contains an invalid placeholder:\n - {}"
+            .format(ve))
+    except KeyError as ke:
+        print_err_and_exit(
+            "Common MIB fragment contains unrecognised placeholder: {}"
+            .format(ke))
+
+    extras_src = read_mib_fragment(extras_file_path)
+
+    full_mib_data = common_src + extras_src
+    write_mib_file(output_mib_path, full_mib_data)
+
+
+if __name__ == "__main__":
+    args = sys.argv[1:]
+
+    # If args are supplied, try to parse the CWC MIB directory.
+    if args:
+        arguments = docopt(__doc__, argv=args)
+
+        # Validate the parsed directory.
+        cwc_dir = os.path.abspath(arguments['--cwc-mib-dir'])
+        full_cwc_fragment_path = os.path.join(cwc_dir, CWC_EXTRAS_NAME)
+        full_cwc_output_mib_path = os.path.join(cwc_dir, CWC_MIB_OUTPUT_NAME)
+
+        if not os.path.isfile(full_cwc_fragment_path):
+            print_err_and_exit(
+                "Supplied cwc-mib-dir does not contain CWC MIB fragment '{}'"
+                .format(CWC_EXTRAS_NAME))
+
+    else:
+        cwc_dir = None
+        sys.stdout.write("No Clearwater Core MIB directory supplied, building "
+                         "Project Clearwater MIB only\n")
+
+    full_pc_output_mib_path = os.path.abspath(PC_MIB_OUTPUT_PATH)
+    full_pc_fragment_path = os.path.abspath(PC_EXTRAS_PATH)
+
+    sys.stdout.write("Generating Project Clearwater MIB at: '{}'\n"
+                     .format(full_pc_output_mib_path))
+
+    generate_mib(full_pc_fragment_path, full_pc_output_mib_path)
```

Comment:
Not sure. I tend to follow the guideline that source files should start at a high level, then introduce lower level concepts as you read on. That way it's easy to pick up the high-level structure of the code, and you only need to look at the low-level detail you are interested in.